core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -86,6 +86,7 @@ private static Block createInternal(int positionCount, Block dictionary, int[] i
                 newIds[position] = dictionaryBlock.getId(ids[position]);
             }
             dictionary = dictionaryBlock.getDictionary();
+            dictionarySourceId = dictionaryBlock.getDictionarySourceId();
```

Comment:
it's still wrong. Consider two blocks with nested dictionaries:

1. 
```
topN:
  sourceId: A
  ids: [0, 2, 2]
    nestedY:
      source: 1
      ids: [1, 2, 3]
      block: X
```

2.
```
topN:
   sourceId: B
   ids: [0, 1]
   nestedV:
      source: 1
      ids: [1, 2, 3]
      block: Y
```

You cannot make them
```
topM':
   sourceId: 1
   ids: [1, 3, 3]
   block: X

topN':
   sourceId: 1
   ids: [1, 2]
   block: X
```

`topM', topN'` cannot have same source ids as `topM'.ids != topN'.ids`. Even if ids were same, they might have originated from different source and just coincidentally be same